### PR TITLE
Use IntLists rather than THLongStorage for sizes/strides for a number of functions in TH.

### PR DESCRIPTION
**Pull request by gchanan**
This changes setStorage (called by set_), newView (view), and newWithStorage (tensor) as per the title.

This also allows us to share size inference code between TH and ATen, so I killed e.g. newInferSize in TH; we just use at::infer_size inside of TH now.
Finally, I reorganized the TH/THC tensor/storage code to more closely match up; e.g. if THC has a non-generic version of a function, TH will as well.

NOTE: I didn't completely kill THSize, THStride yet because I haven't moved all the functions over yet.  This has the unfortunate effect that we now have both IntList and Storage versions of the same concepts, e.g. there are IntListSize and IntListStride in Declarations.cwrap that pull out the LongStorage-ness, but not the 0-dim-to-1-dim or 0-dim-to-null functionality of THLongStorageView.  Both of these limitations will be addressed in further PRs.

### CODE REVIEW

**Comment by gchanan** [comment]
NOTE: this is rebased on top of https://github.com/pytorch/pytorch/pull/10074 -- the right code to review is https://github.com/pytorch/pytorch/pull/10209/files/fde356244d98e4a10a1fd50a67429658b9eae788..HEAD

File: aten/src/ATen/InferSize.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#include <ATen/optional.h>
+#include <ATen/ScalarType.h>
+#include <sstream>
+#include <vector>
+
+namespace at {
+
+// Infers the size of a dim with size -1, if it exists. Also checks that new
+// shape is compatible with the number of elements.
+static std::vector<int64_t> infer_size(IntList shape, int64_t numel) {
```
Comment (ezyang) [inline aten/src/ATen/InferSize.h]: Assumed that this was just cut and paste.

**Review by ezyang — APPROVED** [review]
very nice

File: aten/src/ATen/THSizeStrideCompat.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+// NOTE: these functions are for compatibility into TH functions that takes sizes and strides.
+// We should just write the TH functions that don't require this, but that involves two steps:
+// 1) first class scalar support (for sizes)
+// 2) differentiating between nullptr/non-nullptr strides (the former "infers" strides).
+
+namespace at {
+
+static inline std::vector<int64_t> get_intlist_size_th(IntList sizes) {
+  if (sizes.size() == 0) {
```
Comment (colesbury) [inline aten/src/ATen/THSizeStrideCompat.h]: If you wanted to make this return an `IntList` you could do:

```
static int64_t one = 1;
return IntList(one);
```

Comment (gchanan) [reply aten/src/ATen/THSizeStrideCompat.h]: Seems like a good idea, I'll do it when I can get rid of the stacked diff.

File: aten/src/ATen/InferSize.h (hunk 1 above)
Comment (gchanan) [reply aten/src/ATen/InferSize.h]: yes.

**Comment by gchanan** [comment]
Closing in favor of https://github.com/pytorch/pytorch/pull/10219.

**Pull request closed without merging**